src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -191,197 +371,190 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
             .build());
   }
 
-  private void launchDataNode(SchedulerDriver driver, Offer offer) {
-    launchNode(
-        driver,
-        offer,
-        HDFSConstants.DATA_NODE_ID,
-        Arrays.asList(HDFSConstants.DATA_NODE_ID),
-        HDFSConstants.NODE_EXECUTOR_ID);
-  }
-
-  private void launchInitialJournalNodes(SchedulerDriver driver, Collection<Offer> offers) {
-    for (int i = 0; i < conf.getJournalNodeCount(); i++) {
-      if (offers.size() > 0) {
-        Offer offer = offers.iterator().next();
-        pendingOffers.remove(offer.getId());
-        launchNode(
-            driver,
-            offer,
-            HDFSConstants.JOURNAL_NODE_ID,
-            Arrays.asList(HDFSConstants.JOURNAL_NODE_ID),
-            HDFSConstants.NODE_EXECUTOR_ID);
-      }
+  private boolean tryToLaunchJournalNode(SchedulerDriver driver, Offer offer) {
+    if (offerNotEnoughResources(offer, conf.getJournalNodeCpus(), conf.getJournalNodeHeapSize())) {
+      log.info("Offer does not have enough resources");
+      return false;
     }
-  }
 
-  private void launchInitialNameNodes(SchedulerDriver driver, Collection<Offer> offers) {
-    for (int i = 0; i < HDFSConstants.TOTAL_NAME_NODES; i++) {
-      if (offers.size() > 0) {
-        Offer offer = offers.iterator().next();
-        pendingOffers.remove(offer.getId());
-        launchNode(
-            driver,
-            offer,
-            HDFSConstants.NAME_NODE_ID,
-            Arrays.asList(HDFSConstants.NAME_NODE_ID, HDFSConstants.ZKFC_NODE_ID,
-                HDFSConstants.JOURNAL_NODE_ID),
-            HDFSConstants.NAME_NODE_EXECUTOR_ID);
+    boolean launch = false;
+    List<String> deadJournalNodes = persistentState.getDeadJournalNodes();
+
+    log.info(deadJournalNodes);
+
+    if (deadJournalNodes.isEmpty()) {
+      if (persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Already running journalnode on %s", offer.getHostname()));
+      } else if (persistentState.dataNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Cannot colocate journalnode and datanode on %s",
+            offer.getHostname()));
+      } else {
+        launch = true;
       }
+    } else if (deadJournalNodes.contains(offer.getHostname())) {
```

Comment:
Does this mean that if an entire JN machine goes down, we'll be stuck waiting to relaunch the deadJN until the same hostname comes back up? Obviously we'd prefer to relaunch on that node if possible, but shouldn't JNs be able to be launched elsewhere, as long as there's still a quorum up? Worth a TODO, if not a new issue.